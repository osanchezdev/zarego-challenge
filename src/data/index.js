export const ADMISION_TYPES = [
  'Freshman Student',
  'Transfer Student',
  'Graduate Student',
  'International Student',
];

export const COLLEGES_DATA = [
  'Bethesda University',
  'College of Alameda',
  'California College San Diego',
  'Columbia College',
  'Foothill College',
  'Golden West College',
  'Hartnell College',
  'ICOHS College',
  'Oxnard College',
  'Pitzer College',
];
