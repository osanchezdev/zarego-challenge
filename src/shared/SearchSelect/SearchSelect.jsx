import React, { useRef, useState, useEffect } from 'react';
import { string, bool, func, arrayOf } from 'prop-types';
import styles from './SearchSelect.module.css';
import Input from '../Input/Input';

const SearchSelect = ({
  name,
  label,
  showLabel,
  prefixText,
  error,
  errorMsg,
  controller,
  onSelectListItem,
  data,
}) => {
  const node = useRef();
  const [showList, setShowList] = useState(false);
  const [searchQuery, setSearchQuery] = useState('');
  const [filteredList, setFilteredList] = useState([]);

  const handleChange = (e) => setSearchQuery(e.target.value);
  const handleFocus = () => setShowList(true);
  const handleSelect = (newValue) => {
    setShowList(false);
    onSelectListItem(newValue);
  };

  useEffect(() => {
    const filterResults = data.filter((item) =>
      item.toLowerCase().includes(searchQuery.toLowerCase())
    );
    setFilteredList(filterResults);
  }, [searchQuery]);

  const handleClick = (e) => setShowList(node.current.contains(e.target));

  useEffect(() => {
    document.addEventListener('mousedown', handleClick);
    return () => {
      document.removeEventListener('mousedown', handleClick);
    };
  }, []);

  return (
    <div ref={node} className={styles.search_select__main_wrapper}>
      <div>
        <Input
          placeholder="Type to search"
          name={name}
          label={label}
          showLabel={showLabel}
          controller={controller}
          icon="search"
          prefixText={prefixText}
          error={error}
          errorMsg={errorMsg}
          onChangeParam={handleChange}
          onFocusParam={handleFocus}
        />
      </div>
      <div
        className={`${styles.search_select__list_wrapper} ${
          showList ? styles.show : ''
        }`}
      >
        {showList && (
          <ul className={styles.search_select__list}>
            {filteredList && filteredList.length ? (
              filteredList.map((item) => (
                <li
                  key={item}
                  className={styles.search_select__list_item}
                  value={item}
                  onClick={(e) => handleSelect(e.currentTarget.innerText)}
                >
                  {item}
                </li>
              ))
            ) : (
              <p className={styles.search_select__list_item_empty}>No items</p>
            )}
          </ul>
        )}
      </div>
    </div>
  );
};
SearchSelect.propTypes = {
  name: string.isRequired,
  label: string.isRequired,
  showLabel: bool,
  prefixText: string,
  error: bool,
  errorMsg: string,
  controller: func.isRequired,
  onSelectListItem: func.isRequired,
  data: arrayOf(string).isRequired,
};

SearchSelect.defaultProps = {
  showLabel: true,
  prefixText: null,
  error: false,
  errorMsg: null,
};

export default SearchSelect;
