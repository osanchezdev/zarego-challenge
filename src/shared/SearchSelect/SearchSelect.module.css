.search_select__main_wrapper {
  position: relative;
}

.search_select__list_wrapper {
  opacity: 0;
  position: absolute;
  left: 0;
  top: 75px;
  width: 100%;
  max-height: 170px;
  overflow: auto;
  z-index: 1;
}

.search_select__list_wrapper.show {
  opacity: 1;
}

.search_select__list {
  position: relative;
  margin: 0;
  padding: 0;
  list-style: none;
}

.search_select__list_item {
  cursor: pointer;
  position: relative;
  background: #eee;
  margin: 0;
  padding: 12px;
}

.search_select__list_item:hover {
  background: #ddd;
  transition: background 300ms ease;
}

.search_select__list_item_empty {
  position: relative;
  background: #eee;
  margin: 0;
  padding: 12px;
  text-align: center;
}
