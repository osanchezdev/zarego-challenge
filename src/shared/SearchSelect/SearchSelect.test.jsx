import React from 'react';
import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import SearchSelect from './SearchSelect';

import { COLLEGES_DATA as testData } from '../../data';

describe('SearchSelect component', () => {
  const { labelText, controller, onSelectListItem, placeholder } = {
    placeholder: 'Type to search',
    labelText: 'labelTest',
    controller: {
      ref: () => {},
      onChange: () => {},
      onBlur: () => {},
    },
    onSelectListItem: () => {},
  };
  it('should render component', () => {
    const tree = renderer
      .create(
        <SearchSelect
          label={labelText}
          controller={controller}
          data={testData}
          onSelectListItem={onSelectListItem}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render list items on focus', () => {
    render(
      <SearchSelect
        label={labelText}
        controller={controller}
        data={testData}
        onSelectListItem={onSelectListItem}
      />
    );
    const input = screen.getByPlaceholderText(placeholder);
    fireEvent.focus(input);
    const listItems = screen.queryAllByRole('listitem');
    expect(listItems).toHaveLength(testData.length);
  });

  it('should render no results', async () => {
    render(
      <SearchSelect
        label={labelText}
        controller={controller}
        data={testData}
        onSelectListItem={onSelectListItem}
      />
    );
    const input = screen.getByPlaceholderText(placeholder);
    act(() => {
      userEvent.type(input, 'wasrereqw');
    });
    await waitFor(() => {
      screen.getByText(/No items/i);
    });
    const noResultsItem = screen.getByText(/No items/i);
    expect(noResultsItem).toBeInTheDocument();
  });

  it('should hide list in outside click', async () => {
    render(
      <SearchSelect
        label={labelText}
        controller={controller}
        data={testData}
        onSelectListItem={onSelectListItem}
      />
    );
    const input = screen.getByPlaceholderText(placeholder);
    fireEvent.focus(input);
    let listItems = screen.queryAllByRole('listitem');
    expect(listItems).toHaveLength(testData.length);
    act(() => {
      userEvent.click(document.body);
    });
    listItems = screen.queryAllByRole('listitem');
    expect(listItems).toHaveLength(0);
  });
});
