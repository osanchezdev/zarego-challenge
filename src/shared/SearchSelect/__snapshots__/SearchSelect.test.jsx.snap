// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchSelect component should render component 1`] = `
<div
  className="search_select__main_wrapper"
>
  <div>
    <div
      className="input__main_wrapper"
    >
      <div
        className="input_label__wrapper"
      >
        <label
          htmlFor="nameField"
        >
          labelTest
        </label>
      </div>
      <div
        className="input__wrapper"
      >
        <input
          className="input  input_icon"
          id="nameField"
          name="nameField"
          onBlur={[MockFunction]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Type to search"
          type="text"
        />
        <img
          alt="input-icon"
          className="input_icon_image"
          src="search.svg"
        />
        <div
          className="input__error_msg_wrapper "
        />
      </div>
    </div>
  </div>
  <div
    className="search_select__list_wrapper "
  />
</div>
`;
